### ISSUE TITLE
Make findByCondition() public or allow findOne()/findAll() query condition modification

### ISSUE THREAD

**Issue by vercotux** [yiisoft/yii2#14960]
I cannot even count how many times I've encountered this issue.

I want to call `SomeRecord::findAll($pks)` where `$pks` is an array of primary keys.
However, I just want to pass another condition into that query or set a custom `index()` or different `order()`... But I can't because `findAll()` goes straight to `all()`.

Ok so, let me just call `SomeRecord::findByCondition($pks)->myCondition()->all()` then... Well darn, I can't do that either since it's [`protected`](https://github.com/yiisoft/yii2/blob/277468a7aa7c182ca7a686d3d82a2e38d897a980/framework/db/BaseActiveRecord.php#L127).

Could we make `findByCondition()` public? Either that or add a second `callable $query` parameter to `findAll()` and `findOne()` (similar to how it's done in `Query::with()`)?

**Comment by samdark** [comment]
How about the following?

```php
SomeRecord::find()->where(['id' => $pks])->myCondition()->all()
```

**Comment by vercotux** [comment]
@samdark That doesn't work. You are assuming the PK name is 'id'. It is not, it can be anything.

Can we please reopen this?

**Comment by samdark** [comment]
Making it public may make sense...

**Comment by Alex-Code** [comment]
Would this cause confusion with extra conditions?

e.g.
```php
SomeRecord::findByCondition($pks)->where('true=false')->all()
```
Would override the `where` set by `findByCondition`.

**Comment by vercotux** [comment]
It would of course, but this problem is not unique to `findByCondition()`. Anytime you call `where()` you are overriding any previously set conditions (e.g. a soft delete condition set in activequery `init()`) which is the expected behavior of `where()` (as opposed to `andWhere()`).

We could also move the `findByCondition()` logic into ActiveQuery and call it `primaryKey($key)`? It would be incredibly useful too.

**Comment by rob006** [comment]
@vercotux 
> You are assuming the PK name is 'id'. It is not, it can be anything.

But you should know your PK name, so just replace `id` key with name of PK column.

**Comment by vercotux** [comment]
@rob006 That is not true. There are countless situations where you do not know the PK name, such as creating a behavior or some other form of abstract component that is supposed to work with different models.

**Comment by rob006** [comment]
@vercotux If you don't know PK name you probably don't know anything about table schema. How do you want to add additional conditions, if you don't know what columns this table contains.

**Comment by vercotux** [comment]
@rob006 Oh really? Then what are findAll() and findOne() doing in BaseActiveRecord if they "don't know PK names"? I think you may be confused. When I say I "don't know" the PK name I mean that it is a variable.

**Comment by rob006** [comment]
@vercotux I'm talking about column names, not values. If you don't know which column is used as PK, how could you create any other condition (since you should know nothing about any other column name in this table).
If you can't use `SomeRecord::find()->where(['id' => $pks])->andWhere(['status' => 1])->all()` because "I don't  know that `id` is my PK column", then how could you know that you should use `status` as column name in `andWhere()`?

**Comment by vercotux** [comment]
@rob006 I know you are talking about column names. I never mentioned any values.
And of course you can't add conditions to any unknown columns. Nobody is trying to do that.

You do know which column is used as a PK. See how `findByCondition()` does it.

What you can do is add more conditions to the same PK column. You can also [reindex](http://www.yiiframework.com/doc-2.0/yii-db-querytrait.html#indexBy()-detail) it. You can specify a different [order by](http://www.yiiframework.com/doc-2.0/yii-db-querytrait.html#orderBy()-detail) direction. You can add a [limit](http://www.yiiframework.com/doc-2.0/yii-db-querytrait.html#limit()-detail) etc and plenty of other things. None of this requres you to know the full schema.

**Comment by rob006** [comment]
@vercotux You can't use `indexBy()` or `orderBy()` if you don't know column names. And I can't really imagine any useful usage of `limit()` if you can't set order.

Anyway, I'm against making `findByCondition()` public. We could add something like `andWherePkIs($value)` to `ActiveQuery`, but this is debatable. At least one real use case would be helpful.

**Comment by vercotux** [comment]
@rob006 
>You can't use indexBy() or orderBy() if you don't know column names.

But you *can* know the PK name, so you *can* use these methods. You just need to re-implement exactly what is already implemented in findByCondition() first. Seems pretty silly to demand that for no apparent reason when it could simply be public.

There's tones of use cases. Here's a just few off the top of my head:

```php
SomeRecord::find()->andWherePkIs($pks)->asArray()->all(); // selfexplanatory
SomeRecord::find()->andWherePkIs($pks)->limit(5)->offset(5)->all(); // paginate through $pks that are found
SomeRecord::find()->andWherePkIs($pks)->exists(); // check if any of the $pks exist
SomeRecord::find()->andWherePkIs($pks)->count(); // count how many of the $pks exist
SomeRecord::find()->andWherePkIs($pks)->column();
SomeRecord::find()->andWherePkIs($pks)->each();
```

If there was also a way to invert the condition you could do things like search for all records with an exclusion list:

```php
SomeRecord::find()->andWherePkIsNot($exclude_pks)->all();
```

**Comment by ptz-nerf** [comment]
What about?
```php
SomeRecord::find()->where([SomeRecord::primaryKey()[0] => $pks])->myCondition()->all()
```
I saw no real case to change anything.

**Comment by vercotux** [comment]
@ptz-nerf That violates [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself). You have to keep repeating that same monstrosity everywhere.

**Comment by rob006** [comment]
@vercotux Accessing PK when you don't know it's column name is not a common case. But if you really need it, you can create own helper for that.

**Comment by ptz-nerf** [comment]
@vercotux, i'm neutral to public findByCondition(). But I'm against a bunch of new methods: (and/or)wherePkIs, (and/or)wherePkIsNot... It's redundant.

**Comment by FabrizioCaldarelli** [comment]
Some real example when it is needed dynamic primary key?

**Comment by vercotux** [comment]
@rob006 
>Accessing PK when you don't know it's column name is not a common case. But if you really need it, you can create own helper for that.

Then remove `findOne()` and `findAll()` because you are contradicting yourself.

@ptz-nerf Redundant to *you* because you've never dug deep enough to need that.

@FabrizioCaldarelli I gave you examples. You want me to paste the entire NDA'd project in here?


Also if it's SUCH A HASSLE making a tiny improvement that has no downsides (you just can't see the benefits yourself) then forget it. I'll just make my own improvements. Why waste my time.

**Comment by samdark** [comment]
I want to hear @yiisoft/core-developers and @yiisoft/reviewers opinions on the matter.

**Comment by FabrizioCaldarelli** [comment]
@vercotux I think that having multiple ways to solve a question can confuse the developer. As last solution you can dinamically get primary key from table schema and then apply andWhere() methods to $query object.

**Comment by samdark** [comment]
Yes, I've got it. Waiting for more opinions.

**Comment by vercotux** [comment]
@FabrizioCaldarelli Oh so you are also in favor of removing `findAll()! ` It does the same as `AR::find()->all()` so we better remove it "not to confuse the developer". What a stupid argument.

**Comment by samdark** [comment]
@vercotux, @FabrizioCaldarelli stop it please. Nothing good comes of it. Let's hear other opinions then decide. Thanks.

**Comment by vercotux** [comment]
@samdark My apologies. I'm done.

**Comment by cebe** [comment]
instead of having another method on the activerecord, we could add something like `wherePrimaryKey()` to the ActiveQuery.

**Comment by dynasource** [comment]
Im against such shortcuts. Especially for usecases like this. 
Its very rare & adds confusion.

**Comment by samdark** [comment]
@dynasource how about @cebe proposal?

**Comment by dynasource** [comment]
it's a solution, but tricky. What if there are more PKs?

**Comment by samdark** [comment]
More PKs?

**Comment by dynasource** [comment]
composite PK

**Comment by cebe** [comment]
> composite PK

```php
OrderItem::find()->wherePrimaryKey([
['item_id' => 1, 'order_id' => 1],
['item_id' => 2, 'order_id' => 1],
['item_id' => 3, 'order_id' => 2],
// ...
])->all();
```

**Comment by dynasource** [comment]
-> in which you have to specify the 'names'
-> making it confusing with the `where` method

**Comment by samdark** [comment]
Making it similar, yes. Nothing wrong with that though.

**Comment by cebe** [comment]
for composite Pks the names are necessary also in findOne() and findAll() so there is nothing new.

**Comment by ptz-nerf** [comment]
Are prefix-modifiers planned (or/and/filter)? How can I negate such condition (fairly common)?

**Comment by cebe** [comment]
> Are prefix-modifiers planned (or/and/filter)?

you can add more conditions afterwards using `andWhere()` and `andFilterWhere()`.

> How can I negate such condition (fairly common)?

You mean to select all records except the ones you specify the PK for? Why?

**Comment by ptz-nerf** [comment]
> you can add more conditions afterwards using andWhere() and andFilterWhere().

Conditions with PK? But I don't know its name... The query can already have conditions before I need to add a condition with PK.

> You mean to select all records except the ones you specify the PK for? Why?

Yes. For example to get some similar records excluding current records.
